src/librustc/infer/combine.rs

```diff
@@ -107,13 +109,68 @@ impl<'infcx, 'gcx, 'tcx> InferCtxt<'infcx, 'gcx, 'tcx> {
                 Err(TypeError::Sorts(ty::relate::expected_found(relation, &a, &b)))
             }
 
-
             _ => {
                 ty::relate::super_relate_tys(relation, a, b)
             }
         }
     }
 
+    pub fn super_combine_consts<R>(
+        &self,
+        relation: &mut R,
+        a: &'tcx LazyConst<'tcx>,
+        b: &'tcx LazyConst<'tcx>,
+    ) -> RelateResult<'tcx, &'tcx LazyConst<'tcx>>
+    where
+        R: TypeRelation<'infcx, 'gcx, 'tcx>,
+    {
+        let a_is_expected = relation.a_is_expected();
+
+        if let (&ty::LazyConst::Evaluated(a_eval), &ty::LazyConst::Evaluated(b_eval)) = (a, b) {
+            match (a_eval.val, b_eval.val) {
+                (ConstValue::Infer(InferConst::Var(a_vid)),
+                 ConstValue::Infer(InferConst::Var(b_vid))) => {
+                    self.const_unification_table
+                        .borrow_mut()
+                        .unify_var_var(a_vid, b_vid)
+                        .map_err(|e| const_unification_error(a_is_expected, e))?;
+                    return Ok(a);
+                }
+
+                // All other cases of inference with other variables are errors.
+                (ConstValue::Infer(InferConst::Var(_)), ConstValue::Infer(_)) |
+                (ConstValue::Infer(_), ConstValue::Infer(InferConst::Var(_))) => {
+                    bug!("tried to combine ConstValue::Infer/ConstValue::Infer(InferConst::Var)")
```

Comment:
Well, the other variants of `InferConst` are:

```rust
/// An inference variable for a const, for use in const generics.
#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd,
         Ord, RustcEncodable, RustcDecodable, Hash, HashStable)]
pub enum InferConst<'tcx> {
    /// Infer the value of the const.
    Var(ConstVid<'tcx>),
    /// A fresh const variable. See `infer::freshen` for more details.
    Fresh(u32),
    /// Canonicalized const variable, used only when preparing a trait query.
    Canonical(DebruijnIndex, BoundVar),
}
```

so, yes, this is due to the placeholder stuff. I don't know whether it will get "cleaned up" or not, I guess that depends a bit on what the final shape of our type story looks like, which is something I've been thinking a lot about (I'd like to share the code between rustc and chalk, for example). 

I think overall I'd like to move away from the "zoo" of variants that exist today and have three:

- free variables: inference variables (existentially quantified) and placeholders (universally quantified)
- bound variables (based on debruijn indices): used for canonicalization, `for<'a>` types, etc.

but that's probably a larger conversation than just this PR.

Reply:
I think in short this part of the code seems ok to land as is, it seems comparable to the other cases, and we can refactor them together.